docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+ - If auto-approval of CSRs should be enabled
+
+Other flags that are set unconditionally:
+ - The `BootstrapSigner` and `TokenCleaner` controllers are enabled
+ - `--root-ca-file` to `ca.crt`
+ - `--cluster-signing-cert-file` to `ca.crt`
+ - `--cluster-signing-key-file` to `ca.key`
+ - `--service-account-private-key-file` to `sa.key`
+ - `--use-service-account-credentials` to true
+
+#### Scheduler
+
+kubeadm doesn't set any special flags to the scheduler
+
+Common properties for the control plane components:
+ - Leader election is enabled for both the controller-manager and the scheduler
+ - `HostNetwork: true` is present on all static pods since there is no network configured yet
+ - The yaml manifests are located in `/etc/kubernetes/manifests`
+
+### Addons
+
+#### mark-master
+
+This addon essentially does just this:
+
+```bash
+kubectl taint node ${master_name} node-role.kubernetes.io/master:NoSchedule
+kubectl label node ${master_name} node-role.kubernetes.io/master=""
+```
+
+#### cluster-info
+
+This phase creates the `cluster-info` ConfigMap in the `kube-public` namespace as defined in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md
+ - The `ca.crt` and the address/port of the apiserver is added to the `cluster-info` ConfigMap in the `kubeconfig` key
+ - Exposes the `cluster-info` ConfigMap to unauthenticated users (i.e. users in RBAC group `system:unauthenticated`)
+
+#### self-hosting
+
+Parses the yaml static pod manifests in `/etc/kubernetes/manifests` and converts them to native Deployments and DaemonSets:
+ - TODO
+
+#### kube-proxy
+
+A ServiceAccount for `kube-proxy` is created in the `kube-system` namespace.
+
+Deploy kube-proxy as a DaemonSet:
+ - the credentials (`ca.crt` and `token`) to the master come from the ServiceAccount
+ - the location of the master comes from a ConfigMap
+ - the `kube-proxy` ServiceAccount is bound to the privileges in the `system:node-proxier` ClusterRole
+
+#### kube-dns
+
+A ServiceAccount for `kube-dns` is created in the `kube-system` namespace.
+
+Deploy the kube-dns Deployment and Service:
+ - it's the upstream kube-dns deployment relatively unmodified
+ - the `kube-dns` ServiceAccount is bound to the privileges in the `system:kube-dns` ClusterRole
+
+the TLS Bootstrap role (`system:node-bootstrapper`), so tokens in the group `kubeadm:kubelet-bootstrap` only can access what's required for TLS Bootstrapping
+
+## `kubeadm phase` UX
+
+ - `kubeadm phase certs selfsign`
+   - Inputs:
+     - `--dns-domain=cluster.local`
+     - `--service-cidr=10.96.0.1/12`
+     - `--alt-names=""`
+     - `--cert-dir=/etc/kubernetes/pki`
+   - Outputs (in the `--cert-dir` directory):
+     - `ca.{crt,key}`
+     - `apiserver.{crt,key}`
+     - `apiserver-kubelet-client.{crt,key}`
+     - `sa.{pub,key}`
+     - `front-proxy-ca.{crt,key}`
+     - `front-proxy-client.{crt,key}`
+ - `kubeadm phase kubeconfig clientcert`
```

Comment:
Not sure if offering a generic `kubeconfig clientcert/token` phase can offer the simplest choice for users, because it requires to call this phase 4 times (kubelet,scheduler,controller-manager,admin), each one with very specific parameters.

A draft alternative proposal, might be specializing `kubeconfig` phase with dedicate sub-commands:
- `kubeconfig kubelet` (Inputs: `--api-server`, `--cert-dir`)
- `kubeconfig scheduler` (Inputs: `--api-server`, `--cert-dir`)
- `kubeconfig controller-manager` (Inputs: `--api-server`, `--cert-dir`)
- `kubeconfig user` (Inputs: `--api-server`, `--cert-dir`, `--name`, `--organization`)
All commands should produce a clientCert (default mode) expet if if `--token` flag is provided)

Additionally
- it could be useful to support a `kubeconfig all` (or simply `kubeconfig` without parameters) allowing to generate all 4 default config files in a single action
- especially for `kubeconfig user` it could be useful to to generate also an alternative output format as a `kubectl config set-cluster`, `kubectl config set credentials` commands and separate client files (e.g. useful for admin managing more that one cluster...)